FirebaseFunctionsSwift/Sources/Codable/Callable+Codable.swift

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Foundation
+import FirebaseFunctions
+import FirebaseSharedSwift
+
+extension HTTPSCallable {
+  enum CallableError: Error {
+    case internalError
+  }
+
+  public func call<T: Encodable, U: Decodable>(_ data: T,
+                                               resultAs: U.Type,
+                                               encoder: StructureEncoder = StructureEncoder(),
```

Comment:
I am new to the party here and just started looking at this.

I wonder if you have considered adding this API to `Functions` instead? This would then look as such:

```
let callable = functions.httpsCallable("HelloWorld", encoder: Foo, decoder: Bar);
callable.call(data);
```

This follows the pattern that Firestore uses on the Web and prescribes the schema for all functions call. You can see how we specify a schema for a Firestore collection here: https://firebase.google.com/docs/reference/js/firestore_.collectionreference.md#collectionreferencewithconverter

Reply:
Hi @schmidt-sebastian ,
Thank you for the suggestion - I think that API strikes a really nice balance between ease of use preventing unintended types at the call site.

I'll go ahead and update the PR and close the related discussion thread.

Thanks again, Sebastian!

Reply:
I am not sure I fully understand the last update. I was hoping we could get rid of the "encoder" and "decoder" from `cal()` and just pass it to `HTTPSCallable`, which would then use the same encoder/decoder for every function call. Let me know if I am missing something here.

Reply:
Sorry, I misunderstood your suggestion then. The reference to the Firestore js api where you send a transform along made me think that you meant the encodable and decodable types (since they corresponds to transforms from a structure to an entity).
I can of course bring in the encoder and decoder at this point in the api too.

Reply:
The main idea behind this is that I think the `encoder` and `decoder` are more tied to the Function rather than the Function invocation - and they should be the same for all invocations.

Reply:
Right. I agree that the configuration of the encoder and decoder is as much a part of the function as the data types in use.
The only reason I'd be hesitant to add it is that I am so accustomed to creating new encoders and decoders for every encoding and decoding... Are we guaranteed that the same encoder can be reused - possibly across different threads simultaneously?

Reply:
The encoder and decoder could of course be passed as factory closures defaulting to constructing a new encoder and decoder. But I guess that API is a tiny bit more confusing that just passing an actual encoder/decoder...

Reply:
Ok, looking back at the implementation of StructureEncoder and StructureDecoder, they both create an instance of the type that actually performs the work upon calling `encode` and `decode`.
So I'd say that it's safe to use the same instance from multiple places and even across threads - with the exception of 'configuring' the encoder and decoder. That's not much worse than the responsibility that users of a JSONEncoder and JSONDecoder have today - even though the use case here makes the types 'live' a bit longer than they might otherwise do...

What do you think?

Reply:
https://firebase.google.com/docs/reference/swift/firebasefirestore/api/reference/Classes/Query.html#endat:

Reply:
I certainly do not have any idiomatic Swift experience so I cannot judge what feels better for the platform. I personally prefer configuring the callable since it might lead to fewer instances where the user that pass an encode (granted, most callables will likely only be invoked at one callsite). 

If this is idiomatic enough, then we can probably just document that the serializer has to be thread-safe and update the PR. The final PR has to go through our internal API review process and we can raise this concern.

Reply:
But the serializer is always an instance of the StructureEncoder type - which is owned by this code base. The encoder is not the precise transform between a specific model entity and it's serialized format - it's a general purpose thing that encodes models to serialized formats. The only reason that it is supplied as an optional argument is that you may wish to configure a few properties of the encoder. The -type- of the encodable entity is closer to representing the transform (through the Codable system and a bit of compiler magic).

Your comment about documenting that the serializer must be thread safe leads me to think that you may be thinking that it's a user supplied transform - like the Firestore js APIs you referred to.

Reply:
I think my understanding is the same as yours. The user will likely not specify their own encoder, but they can if they choose to. So for most users the decision on where to configure the encoder will be irrelevant.

I will create an API proposal and discuss this with our iOS group. I don't think we need to do much here before this gets approved and we can adapt the PR as per the guidelines we receive.

Reply:
Excellent, sorry for misinterpreting - I just wanted to make sure. Yeah, you could of course document that it's not thread safe to modify the configuration of the encoder... Looking forward to seeing what you end up agreeing on. ❤️

Reply:
Thank you for your patience :) I have created the internal proposal and it is pending feedback.

Reply:
No worries, I am in no rush. 😊

I do really like your idea about moving the encoder and decoder arguments from `call` to the `httpsCallable` function in order to make the call site even simpler and configuration more consistent. So regarding the lifetime of the encoder/decoder I see three options:

1. Document that they will be long lived and that you should avoid modifying their state from multiple threads. 

2. Configure them by closures returning an encoder, so that new encoders will be used for each invocation. This would require a bit more explanation in the documentation.

3. Create a thin struct wrapper around `StructureEncoder` named `Functions.Encoder` that Carrie's the same configuration but since it's a value type, it's state can't be modified externally. And ditto for decoder of course.

Do you have a favorite of the three?

Reply:
I am in favour of configuring the encoder / decoder on httpsCallable. 

Morten - you mentioned "modifying the state of the encoder/decoder" - when would a user do this?
Also, I guess thread-safety is a concern for any code that uses an encoder/decoder, so telling the users that if they plan to re-configure the encoder/decoder instance, they need to do so in a thread-safe way. As Sebastian mentioned, most users should be fine with the defaults (at least we should try to make sure they can), so anyone using a custom encoder/decoder should hopefully know what they're getting themselves into :-)

Reply:
Hi @peterfriese ,
Likely they would never do that - although `userInfo` could be used together with an encodable or decodable entity for some custom purpose. 

And you're right that this basically the same as other usages of encoders and decoders - but there is one difference:
The `httpCallable` is specifically intended for longer durations where you create you callable and invoke it many times, where with all other usages I have seen myself, it is customary to create and configure the encoder or decoder for a single use - as in the Codable overlay for Firestore.

It's likely not an issue, but as I mentioned somewhere else in this thread it represents a pattern of usage of the encoder and decoder that I haven't met before.

Reply:
But using something besides the defaults may not be very uncommon. For instance through a preference for snake cased keys in JS, but not in Swift...

Reply:
And apologies for using 'you' to refer to you all working on the Firebase Apple sdk collectively. I know that you may of course have different opinions. 😊

I you wish to try out the two flavors of generics discussed above, toggle between head and head minus two commits. 😊

Reply:
Also note that I am in the process of adding tests so you can see the call site after @paulb777 made it possible to use the required 'fakes' from Swift. I think I'll have the tests ready tomorrow.